lint/mypy.bzl

```diff
@@ -0,0 +1,231 @@
+    # effectively the presence of a cache file prevents reading of the source file
+    # (though the source file must still exist).
+    args.add("--bazel")
+
+    # Undocumented flag, see https://github.com/python/mypy/pull/4759
+    # The --cache-map flag specifies a mapping from source files to cache files that overrides the
+    # usual way of finding the locations for cache files; e.g.
+    # mypy --cache-map foo.py foo.meta.json foo.data.json -- foo.py
+    # There must be any number of triples (source, meta, data) with the additional constraints that
+    # the source must end in .py or .pyi, meta must end in .meta.json,
+    # and data must end in .data.json. For files in the cache map, the default cache directory
+    # (--cache-dir) is ignored, and if the cache files named in the --cache-map don't exist
+    # they will be written. For files not in the cache map, the --cache-dir is still used.
+    args.add("--cache-map")
+
+    # TODO: also need to construct the cache-map for .py files from deps, as seen from --verbose:
+    # LOG:  Could not load cache for src.py_types.testing_deps.foo.fizz: src/py_types/testing_deps/foo/fizz.meta.json
+    # LOG:  Metadata not found for src.py_types.testing_deps.foo.fizz
+    for src in srcs:
+        args.add(src)
+
+        # NB: handle //path/to/package:subdir/file.py -> subdir/file.py
+        path = paths.relativize(src.path, ctx.label.package)
+        for kind in ("meta", "data"):
+            # Note, this assumes every source file has a single mypy aspect producing these output files
+            # For subdir/file.py, declare subdir/file.meta.json
+            # If this is a problem, we could re-root these outputs beneath a target-specific tree
+            file = ctx.actions.declare_file(paths.replace_extension(path, ".%s.json" % kind))
+            cache_map_outputs.append(file)
+            args.add(file)
+
+    # Without this option, mypy will print on success, which will spam the Bazel output:
+    # Success: no issues found in 1 source file
+    args.add("--no-error-summary")
+
+    # Enable with --aspects_parameters=rules_lint_verbose=True
+    if ctx.attr.rules_lint_verbose:
+        args.add("--verbose")
+
+    # TODO: need some of these? Seen at RH
+    # COVERAGE_FLAGS=(
+    #     --linecount-report $report_output_dir
+    #     --linecoverage-report $report_output_dir
+    #     --lineprecision-report $report_output_dir
+    #     --any-exprs-report $report_output_dir
+    # )
+
+    ctx.actions.run(
+        inputs = inputs,
+        outputs = cache_map_outputs,
+        executable = executable,
+        arguments = [args],
+        env = {
+            # TODO: investigate a more robust way to construct this, and avoid the length of the
+            # value overflowing the operating system limit.
+            "PYTHONPATH": ":".join(["external/" + t for t in transitive_imports.to_list()]),
```

Comment:
Does this clobber over the binaries own `PYTHONPATH`? What happens when mypy has plugins?

Reply:
No idea about plugins, will add to the TODO list.

As for how the rules_python runtime stub handles it, it seems to merge the entries:

```
old_python_path = os.environ.get('PYTHONPATH')
  if old_python_path:
    python_path_entries += old_python_path.split(os.pathsep)
```

Reply:
Anyhow I'd like to delete this line altogether if we can make a venv to run inside instead.

Reply:
the other relevant var I saw was MYPYPATH, I forget the difference at this point but maybe that's useful to avoid confusion/clobbering here?

Reply:
@mattem tracked that down - files on the MYPYPATH will be type-checked, in this case we just want the file available for the checker to locate it but want to avoid type-checking 3p code